android_vs_ios.md

```diff
@@ -0,0 +1,175 @@
+
+Data class’ main feature is to provide a default `equals()`, `hashcode()`, and `copy()` method and [destructuring](https://kotlinlang.org/docs/reference/multi-declarations.html). So it is still a class, but with a couple of functions already implemented.
+
+Struct’s main feature to is to serve as value type and enforce immutability, so it behaves differently from class, particularity when the object is passed around. (Swift 4.1 also added automatic synthesizer to implement Equatable and Hashable)
+
+They both forbid subclassing and are often be used for models.
+
+## Extensions + Protocol/Interface
+Swift’s [extension](https://docs.swift.org/swift-book/LanguageGuide/Extensions.html) is a bit more powerful as it allows adding protocol conformance to existing classes (including unmodifiable classes from framework/libraries). So in addition to adding new helper methods to classes, Swift’s extension is also commonly used to provide default implementation of protocols (since protocols doesn’t directly support it). Along with it, the common pattern is to use [protocols+extensions for inheritance](https://medium.com/nsistanbul/protocol-oriented-programming-in-swift-ad4a647daae2) instead a single base class (this is also used to emulate an abstract class).
+
+Kotlin’s [extensions](https://kotlinlang.org/docs/reference/extensions.html) are on a per method basis, so it can only be used to add helper methods and cannot be used to add interface conformance. On the plus side, the extensions can be declared inside another class, which allows it to access the private properties/methods of that class.
+
+Both Kotlin and Java provides [abstract class](https://kotlinlang.org/docs/reference/classes.html#abstract-classes). This, coupled with the (historic) lack of flexibility on interfaces and extensions, has led to many BaseClasses for sharing common functionalities. Kotlin and newer version of Java allows for default implementation within interfaces. However, since it’s kinda new, the pattern of using interfaces to implement common functionality over base class is not well adopted yet.
+
+## Generics
+Java and Kotlin generic has [type erasure](https://stackoverflow.com/a/339708/4051916) (mainly to provide backwards compatibility), which causes the generic's type info to be lost at runtime and limits its usability. Kotlin inline functions does allow reified generic to provide generic type info at runtime.
+
+In Swift, the generic's type info is available at runtime. So you can get do class check, call its init, etc from the generic class.
+
+## Enum
+Swift’s enum is much more powerful than Java/Kotlin’s enum as they can hold [associated values](https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html). However, Kotlin’s [sealed class](https://kotlinlang.org/docs/reference/sealed-classes.html) basically function the same way as Swift enums.
+
+## Other Kotlin features
+* `if` and `when` statements can also return value. ex: `val z = if (x>y) 5 else 10`
```

Comment:
Also `try`, though maybe that's similar in Swift so you didn't want to call it out here